cmd/kubeadm/app/util/runtime.go

```diff
@@ -0,0 +1,153 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"strings"
+
+	kubeadmapiv1alpha2 "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha2"
+	utilsexec "k8s.io/utils/exec"
+)
+
+type containerRuntimeType string
+
+// Supported runtime types
+const (
+	Docker containerRuntimeType = "docker"
+	CRI    containerRuntimeType = "cri"
+)
+
+// ContainerRuntime is a struct that interfaces with the container runtime
+type ContainerRuntime struct {
+	runtimeType containerRuntimeType
+	exec        utilsexec.Interface
+	tool        string
+	args        []string
+}
+
+// NewContainerRuntime sets up and returns a ContainerRuntime struct
+func NewContainerRuntime(execer utilsexec.Interface, criSocket string) (*ContainerRuntime, error) {
+	var err error
+	var toolName string
+	runtime := ContainerRuntime{exec: execer}
+
+	if criSocket != kubeadmapiv1alpha2.DefaultCRISocket {
+		runtime.runtimeType = CRI
+		runtime.args = []string{"-r", criSocket}
+		toolName = "crictl"
+	} else {
+		runtime.runtimeType = Docker
+		toolName = "docker"
+	}
+
+	if runtime.tool, err = execer.LookPath(toolName); err != nil {
+		return nil, fmt.Errorf("%s is required for %s container runtime: %v", toolName, runtime.runtimeType, err)
+	}
+
+	return &runtime, nil
+}
+
+// IsDocker returns true if the runtime is docker
+func (runtime *ContainerRuntime) IsDocker() bool {
+	return runtime.runtimeType == Docker
+}
+
+// IsRunning checks if runtime is running
+func (runtime *ContainerRuntime) IsRunning() error {
+	if err := runtime.Run("info"); err != nil {
+		return fmt.Errorf("container runtime is not running: %v", err)
+	}
+	return nil
+}
+
+func (runtime *ContainerRuntime) getCommand(args ...string) utilsexec.Cmd {
+	cmdArgs := append(runtime.args, args...)
+	return runtime.exec.Command(runtime.tool, cmdArgs...)
+}
+
+// Run creates command object and runs it to the completion
+func (runtime *ContainerRuntime) Run(args ...string) error {
+	return runtime.getCommand(args...).Run()
+}
+
+// CombinedOutput runs the command and returns its combined standard output
+// and standard error. This follows the pattern of package os/exec.
+func (runtime *ContainerRuntime) CombinedOutput(args ...string) ([]byte, error) {
+	return runtime.getCommand(args...).CombinedOutput()
+}
+
+// ListContainers lists running containers
+func (runtime *ContainerRuntime) ListContainers() ([]string, error) {
+	var output []byte
+	var err error
+
+	switch runtime.runtimeType {
```

Comment:
What do you think about having one case statement in the constructor that creates some new struct (DockerInterface or CrictlInterface) that implements an interface which contains the CRI functions needed for kubeadm?

This would reduce the case statements in each function and call the Interface instead. This makes it easier to mock in tests and pushes the implementation out of the ContainerRuntime to be isolated in some struct. Adding new runtimes would also be easier but I admit that will likely not happen any time soon.

These functions could shrink down to `return runtime.CRI.List()`, `return runtime.CRI.Rm(containers...)` or something similar.

Reply:
@chuckha Will do. Thank you for the review!